packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +57,103 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir, assembly);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string, assembly: CloudAssembly) {
+  const lazyHash = new LazyHash(outdir);
+
+  const templatePathsByPlugin: Map<IPolicyValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...assembly.stacks.map(stack => stack.templateFullPath));
```

Comment:
The `root` that is passed to this function will either be the `App` or an individual `Stage` (synthesis is called separately for each Stage) so it will only get the stacks under that stage.